### ISSUE TITLE
cannot use fat copy on some files

### ISSUE THREAD

**Issue by steve6375** [chenall/grub4dos#115]
[FATCOPY_ISS115.zip](https://github.com/chenall/grub4dos/files/216839/FATCOPY_ISS115.zip)

I have noticed that 'fat copy' command does not copy some files but has no problem with many other similar files.
A test case is attached

**Comment by steve6375** [comment]
![image](https://cloud.githubusercontent.com/assets/1469350/14490499/19a0188e-016b-11e6-96fa-14faa20988e8.png)

**Comment by steve6375** [comment]
Please point me at latest version of fat executable?
Seems to work OK if the same file is on the USB drive (not in ISO)
(bd)/fat copy (bd)/ocmanage.dll (fd0)/ || pause copy ocmanage.dll failed!
is OK. So problem with reading ISO???

**Comment by steve6375** [comment]
I compiled latest version of fat and get the same problem.

**Comment by steve6375** [comment]
Same issue with 20160413

**Comment by steve6375** [comment]
Using this modified code in copyFF.g4b, it proves that the problem is nothing to do with the fat utility, because dd also fails on same fails that fat copy fails on (dummy.txt is just a large file). So the problem is reading certain files from inside the ISO. Some are OK but others are not!

```
::copy file
echo FILE=%FD%%%cd%/%1
debug 1
#dd if=(bd)/dummy.txt of=(bd)/dummy.txt || echo Error: dd failed on dummy test
dd if=()/%1 of=(bd)/dummy.txt > nul || echo **** ERROR dd failed on  %cd%/%1 - dd if=()/%1 of=(bd)/dummy.txt
fat copy ()/%1 %FD%%%cd%/ > nul || echo **** ERROR FAT %cd%/%1
echo -----------------------------
pause
shift
goto :copyfiles

```

![image](https://cloud.githubusercontent.com/assets/1469350/14581487/8d603588-03e6-11e6-86bc-204c0c47f89f.png)

**Comment by 2012yaya2007** [comment]
Error using the "DD" command. 
"DD" is not possible to create a file.

**Comment by steve6375** [comment]
dummy.txt already exists. You can see it works for ocmanager.dll and setup.exe, but not for some other files.

**Comment by steve6375** [comment]
Any progress please? This seems to be a bug in grub4dos?  Can you reproduce the issue using the test case I uploaded?

**Comment by 2012yaya2007** [comment]
Fat may have problems. DD no problem.

::copy file
echo  dd if=()%%cd%/%1 of=()/1.txt
dd if=()%%cd%/%1 of=()/dummy.txt > nul || pause   ERROR  - dd if=()%%cd%/%1 of=()/dummy.txt
shift
goto :copyfiles

**Comment by steve6375** [comment]
Yes sorry bad code
It is just fat problem

**Comment by steve6375** [comment]
Yes sorry bad code
It is just fat problem but if copy same file from (hd0,0)/ocmanage.dll then it works. So problem is when copying file from inside an ISO???

**Comment by steve6375** [comment]
If I rename ocmanage.dll to rcmanage.dll then works.  If I rename setup.exe to ocmanage.dll inside ISO then it fails. So it is the file name that has a problem.

**Comment by 2012yaya2007** [comment]
FAT no problem. 

::copy file
echo  fat copy ()%cd%/%1 %FD%%%cd%/
fat copy ()%cd%/%1 %FD%%%cd%/ > nul || pause Error: copy ()%cd%/%1 %FD%%%cd%/
shift
goto :copyfiles

**Comment by 2012yaya2007** [comment]
Fd0 to be large enough to accommodate the contents of TEST.iso

**Comment by steve6375** [comment]
Try this code: it just copies the same file in two different ways
why does it work for some files but not other flles???
I test for full path after copy and delete file before 2nd copy.

```
set C1=
set C2=
fat copy ()%cd%/%1 %FD%%%cd%/ > nul || set C1=1
if exist C1 echo ERROR1: copy ()%cd%/%1 %FD%%%cd%/ || echo OK1  : copy ()%cd%/%1 %FD%%%cd%/ 
if not exist %FD%%%cd%/%1 echo COPY1 FAILED

fat del %FD%%%cd%/%1 > nul

fat copy /%1 %FD%%%cd%/ || set C2=1
if exist C2 echo ERROR2: %cd%/%1 || echo OK2  : copy /%1 %FD%%%cd%/
if not exist %FD%%%cd%/%1 echo COPY2 FAILED

shift
goto :copyfiles

```

![image](https://cloud.githubusercontent.com/assets/1469350/14937608/b21506e6-0f04-11e6-94bf-decae0166d5d.png)

**Comment by 2012yaya2007** [comment]
This example would like to explain what? 
1. fat copy ()%cd%/%1 %FD%%%cd%/ > nul || set C1=1 
  Copy oracle.cab success .
2. You delete the oracle.cab.
3. fat copy /%1 %FD%%%cd%/ || set C2=1
  Copy oracle.cab fail. Tell you that oracle.cab does not exist.

Don't forget, the previous copy of the successful oracle.cab , has been deleted by you ,

**Comment by steve6375** [comment]
If you look at the screenshot, you can see that many files are copied OK (preload.exe, etc. etc.). Only certain files fail to copy.

The root is set to the /add-ons/pws directory, so both forms of the 'fat copy' command should work (and do work for many files in same directory - but not all files!).
You can see that both 'fat copy' commands work for many files, but not for some files e.g. ocmanage.dll and oracle.cab.
I have proved that it is the file **name** that is the problem, if I rename ocmanage,dll to setup.exe inside the ISO, and rename setup.exe to ocmanage.dll, then  ocmanage.dll still fails (which is really setup.exe). So it is the file name that is causing the problem.

If I rename ocmanage.dll to **r**cmanage.dll (inside the ISO and save new ISO) then works. So it is not the length of the file name. If I rename rcmanage.dll back to ocmanage.dll and re-save the ISO, then it fails again!

**Comment by 2012yaya2007** [comment]
How FAT works, I don't know. 
however :
fat copy ()%cd%/%1 %FD%%%cd%/   OK!
 fat copy /%1 %FD%%%cd%/  fail!

**Comment by 2012yaya2007** [comment]
FAT has a parameter '/o' ,Problem by it .
use:
1. fat copy ()%cd%/%1 %FD%%%cd%/
2. fat copy //%1 %FD%%%cd%/

**Comment by steve6375** [comment]
Yes! I can't believe I didn't spot that only files beginning with o had a problem!
So it is a bug in fat.c. 
    if (memcmp(arg, "/o", 2) == 0)
change to...
    if (memcmp(arg, " /o ", 4) == 0)

Then it will only fail if there is a file called o or is there a better fix?

**Comment by steve6375** [comment]
There are also other bugs in fat.c
		else if (memcmp (arg, "dir", 3) == 0)

For instance, if I have a file called direction.txt then fat will not copy it

e.g.

fat copy /direction.txt (fd0)/

![image](https://cloud.githubusercontent.com/assets/1469350/14953109/4262fb36-105e-11e6-9e15-c6f90e46ec32.png)

**Comment by steve6375** [comment]
Can you fix the fat.c code and recompile please?
